### PR TITLE
Integrate WavPack library to export/import wavpack files

### PR DESCRIPTION
**Pull request by Subhra264**
Resolves: #1470 

*(short description of the changes and the motivation to make the changes)*
This Pull Request is created to integrate wavpack with audacity so that users can import/export wavpack files.

### Export:
1. By default, hybrid mode is not selected and the wavpack library creates a pure lossless "wv" file. If hybrid mode is enabled and the "created correction file" option is not selected then it will create a single lossy "wv" file.
Export wavpack image 1

2.  Enabled "create correction file", so it will create a lossy "wv" file along with its "wvc" correction file.
Export wavpack image 2

3. Added the value for "Artist Name" tag.
Export wavpack image 3

4. 
Export wavpack image 4

5. 
Export wavpack image 5 (with wvc)

### Import
1. 
Import wavpack image 1(with wvc)

2. 
Import wavpack image 2(with wvc)

3.
Import wavpack image 3(with wvc)

4. Exported tags are present in the imported audio
Import wavpack image 4(with wvc)

Also, the exported "wv" file successfully opens with the "VLC Media Player".

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [ ] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

**Comment by LWinterberg** [comment]
Hello @Subhra264, are you planning to continue work on this?

**Comment by Subhra264** [comment]
Hello, sorry, I know that I am late( I had no idea about audio channels, samples, sample rate etc. so had to study it first), but I am still working on it..

The `ImportWavPack.cpp` file is almost ready, except the `ImportWavPackFileHandle::Import` function which is partly complete. For example I pushed all the `channels` to `outTracks`. But it's taking time because I am not sure what to do with the `samplesRead` decoded with `WavpackUnpackSamples` as  in  the `ImportFLAC.cpp` file, I saw that channels were directly pushed to `outTracks`,  but in `ImportOgg.cpp` file, the read samples were appended to channels.
Also in `Import.cpp` and `ProjectFileManager.cpp`(I forgot the exact name but it was something similar), it looked like only those `tracks` were used.
Basically, I am confused that if only channels created with `WaveTrackFactory` and `sampleRate` are enough or I need to somehow append the decoded samples to the channels.

Hopefully I will be able complete in 1 or 2 days.
Thanks...

**Comment by crsib** [comment]
Hi @Subhra264 

Sorry for the delay! Can you please rebase onto the latest master, squash you changes into a single commit and force push?

**Comment by Paul-Licameli** [comment]
> Hi @Subhra264
> 
> Sorry for the delay! Can you please rebase onto the latest master, squash you changes into a single commit and force push?

I would like to see a rebase too, but not necessarily a squash to one commit, if the intermediate steps are really useful for review of the new feature in sensible incremental steps, and not just corrections of mistakes.

**Comment by Subhra264** [comment]
> Sorry for the delay!

No Issues! :)
>  but not necessarily a squash to one commit, if the intermediate steps are really useful for review of the new feature in sensible incremental steps, and not just corrections of mistakes.

yeah, actually, I added new commits after the previous commit is bug-free ( I mean, after resolving the bugs I encountered related to that particular feature/implementation, so may not be "bug-free" ). So, each commit mostly represents a new feature/file. For example, the first commit is about integrating wavpack and creating a file containing all the necessary class details, the second commit implements all the `ImportPlugin` and `ImportFileHandle` class methods, and the third commit is to read tags, etc.

**Comment by Subhra264** [comment]
> I have crafted a commit, that should resolve your CI issues https://github.com/crsib/audacity/commit/2a353fd357f6cebb1a9cda61be33a9afe3edfb36

All checks passed! Thank you!

File: src/export/ExportWavPack.cpp [hunk 1]
```diff
@@ -0,0 +1,431 @@
+/**********************************************************************
+
+   Audacity: A Digital Audio Editor
+
+   ExportWavPack.cpp
+
+   Subhradeep Chakraborty
+
+   This program is distributed under the GNU General Public License, version 2.
```
Comment (crsib) [inline src/export/ExportWavPack.cpp:9]: Please update to GPL2-or-later.

Or even better, please add `/*  SPDX-License-Identifier: GPL-2.0-or-later */` to the top of both Export and Import

Comment (Subhra264) [reply src/export/ExportWavPack.cpp]: done

File: src/export/ExportWavPack.cpp [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+#define SAMPLES_PER_RUN 8192u
+
+   uint32_t bufferSize = SAMPLES_PER_RUN * numChannels;
+   ArrayOf<int32_t> wavpackBuffer{ bufferSize };
+   auto updateResult = ProgressResult::Success;
+   {
+      auto mixer = CreateMixer(tracks, selectionOnly,
+         t0, t1,
+         numChannels, SAMPLES_PER_RUN, true,
+         rate, int16Sample, mixerSpec);
```
[the remark this answers is not in the document]
Comment (crsib) [reply src/export/ExportWavPack.cpp:328]: (I believe, that Audacity will return you an array of 32 ints, if you request sample24 as an output format for the mixer. Am I right, @Paul-Licameli?)

File: src/export/ExportWavPack.cpp [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+
+      while (updateResult == ProgressResult::Success) {
+         auto samplesThisRun = mixer->Process(SAMPLES_PER_RUN);
+
+         if (samplesThisRun == 0)
+            break;
+
+         char *mixed = (char *)(const char*)mixer->GetBuffer();
+         for (decltype(samplesThisRun) j = 0; j < samplesThisRun; j++) {
+            for (size_t i = 0; i < numChannels; i++) {
```
Comment (crsib) [inline src/export/ExportWavPack.cpp:344]: This cycle should be adopted to the different sample formats

File: src/export/ExportWavPack.cpp [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+               int32_t value = *mixed++ & 0xff;
+               value += *mixed++ << 8;
+               wavpackBuffer[j*numChannels + i] = value;
+            }
+         }
+
+         if (!WavpackPackSamples(wpc, wavpackBuffer.get(), samplesThisRun))
+            updateResult = ProgressResult::Cancelled;
```
Comment (crsib) [inline src/export/ExportWavPack.cpp:352]: This can be `return ProgressResult::Failed` now, as memory managment is handled by C++ now

File: src/export/ExportWavPack.cpp [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+
+         if (updateResult == ProgressResult::Success)
+            updateResult =
+               progress.Update(mixer->MixGetCurrentTime() - t0, t1 - t0);
+      }
+   }
+
+   if (!WavpackFlushSamples(wpc)) {
+      updateResult = ProgressResult::Cancelled;
```
Comment (crsib) [inline src/export/ExportWavPack.cpp:361]: `return ProgressResult::Failed`?

And I think that `BasicUI::ShowErrorDialog` should be used to notify the user about the error in places where codec has failed. 

(Please, see ExportPCM for example)

File: src/export/ExportWavPack.cpp [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+
+      wxString n;
+      for (const auto &pair : metadata->GetRange()) {
+         n = pair.first;
+         const auto &v = pair.second;
+         if (n == TAG_YEAR) {
+            n = wxT("DATE");
+         }
+         WavpackAppendTagItem(wpc,
+                              (char *) (const char *) n.mb_str(wxConvUTF8),
```
Comment (crsib) [inline src/export/ExportWavPack.cpp:374]: Both conversions are redundant

File: src/export/ExportWavPack.cpp [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+                              (char *) (const char *) v.mb_str(wxConvUTF8),
```
Comment (crsib) [inline src/export/ExportWavPack.cpp:375]: Ditto

File: src/export/ExportWavPack.cpp [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+                              (int) v.length());
+      }
+
+      if (!WavpackWriteTag(wpc)) {
+         // Not sure what to do
+      }
+   }
+
+   WavpackCloseFile(wpc);
```
Comment (crsib) [inline src/export/ExportWavPack.cpp:384]: No longer needed!

File: src/export/ExportWavPack.cpp [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+   }
+
+   return updateResult;
+}
+
+// Based on the implementation of write_block in dbry/WavPack
+// src: https://github.com/dbry/WavPack/blob/master/cli/wavpack.c
+int ExportWavPack::WriteBlock(void *id, void *data, int32_t length)
+{
+   WriteId *outId = (WriteId *) id;
```
Comment (crsib) [inline src/export/ExportWavPack.cpp:398]: `static_cast`

File: src/export/ExportWavPack.cpp [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+
+   uint32_t bcount;
```
Comment (crsib) [inline src/export/ExportWavPack.cpp:400]: Unused!

File: src/export/ExportWavPack.cpp [hunk 11]
```diff
@@ -0,0 +1,430 @@
...
+
+   if (outId->error)
+      return FALSE;
```
Comment (crsib) [inline src/export/ExportWavPack.cpp:403]: `return false;`

File: src/export/ExportWavPack.cpp [hunk 12]
```diff
@@ -0,0 +1,430 @@
...
+
+   if (outId && outId->file && outId->file.get() && data && length) {
```
Comment (crsib) [inline src/export/ExportWavPack.cpp:405]: This function can be rewritten with less nesting involved!

```
int ExportWavPack::WriteBlock(void *id, void *data, int32_t length)
{
    if (id == nullptr || data == nullptr || length == 0)
        return true; // This is considered to be success in wavpack.c reference code

    WriteId *outId = static_cast<WriteId*>(id);

    if (!outId->file)
        // This does not match the wavpack.c but in our case if file is nullptr - 
        // the stream error has occured
        return false; 

    if (!outId->file->Write(data, length).IsOk()) {
        outId->file.reset();
        return false;
    }

    outId->bytesWritten += length;

    if (outId->firstBlockSize == 0)
        outId->firstBlockSize = length;

    return true;
}
```

Also, this allows to remove `error` from the WriteId structure!

**Review by crsib — CHANGES_REQUESTED** [review]
When I export from Audacity - the resulting file seems to be missing some metadata: VLC does not show the file length and does not allow seeking. Everything works fine, if I generate wv file using the official CLI.

Looks like you miss `WavpackUpdateNumSamples` call in the end.

This concludes the review of the ExportWavPack.cpp, review for the importer will follow.

**Review by crsib — CHANGES_REQUESTED** [review]
When I export from Audacity - the resulting file seems to be missing some metadata: VLC does not show the file length and does not allow seeking. Everything works fine, if I generate wv file using the official CLI.

Looks like you miss `WavpackUpdateNumSamples` call in the end.

This concludes the review of the ExportWavPack.cpp, review for the importer will follow.

**Comment by Subhra264** [comment]
> When I export from Audacity - the resulting file seems to be missing some metadata: VLC does not show the file length and does not allow seeking. Everything works fine, if I generate wv file using the official CLI.
> 
> Looks like you miss `WavpackUpdateNumSamples` call in the end.

I have replaced `FileIO` with `wxFile` as seeking to the first block of the file was required. I have tested the wavpack files generated from audacity with VLC media player. Now, it seems to work correctly as it is now allowing seeking and the audio length is also shown. Also, I have addressed the changes you asked me to make. The latest commit is now ready for another review.

File: src/export/ExportWavPack.cpp [hunk 13]
```diff
@@ -371,21 +413,39 @@ ProgressResult ExportWavPack::Export(AudacityProject *project,
...
-
    if ( !outWvFile.file.get()->Close()
       || ( outWvcFile.file && outWvcFile.file.get() && !outWvcFile.file.get()->Close())) {
-      updateResult = ProgressResult::Cancelled;
+      return ProgressResult::Failed;
+   }
+
+   // wxFile::Create opens the file with only write access
+   // So, need to open the file again with both read and write access
+   if (!outWvFile.file->Open(fName.GetFullPath(), wxFile::read_write)) {
```
Comment (crsib) [inline src/export/ExportWavPack.cpp:434]: For simplicity we can cache that block in memory, but this will do too. Very good!

File: src/export/ExportWavPack.cpp [hunk 14]
```diff
@@ -425,6 +486,4 @@ void ExportWavPack::OptionsCreate(ShuttleGui &S, int format)
 
 static Exporter::RegisteredExportPlugin sRegisteredPlugin{ "WavPack",
    []{ return std::make_unique< ExportWavPack >(); }
-};
-
-#endif
+};
```
Comment (crsib) [inline src/export/ExportWavPack.cpp:489]: We generally prefer to have an empty new line at the end of the file

File: src/export/ExportWavPack.cpp [hunk 15]
```diff
@@ -0,0 +1,489 @@
...
+{
+   return true;
+}
+
+bool ExportWavPackOptions::TransferDataFromWindow()
+{
+   ShuttleGui S(this, eIsSavingToPrefs);
+   PopulateOrExchange(S);
+
+   CreateCorrectionFileSetting.Write( mCreateCorrectionFile->GetValue() );
```
Comment (crsib) [inline src/export/ExportWavPack.cpp:217]: Hmm, I'm not sure why this is needed. 

And I think we need to flush settings here

**Review by crsib — COMMENTED** [review]
Export looks fine now, I will review import now!

**Comment by Subhra264** [comment]
> Export looks fine now, I will review import now!

Ok then, I will commit and push the remaining ExportWavPack changes along with ImportWavPack changes.

File: src/import/ImportWavPack.cpp [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+
+            // Get the actual length of the item key at this index i
+            itemLen = WavpackGetTagItemIndexed(mWavPackContext, i, NULL, 0);
+            item = (char *)malloc(itemLen + 1);
+            WavpackGetTagItemIndexed(mWavPackContext, i, item, itemLen + 1);
+            name = UTF8CTOWX(item);
+
+            // Get the actual length of the value for this item key
+            valueLen = WavpackGetTagItem(mWavPackContext, item, NULL, 0);
+            itemValue = (char *)malloc(valueLen + 1);
```
Comment (crsib) [inline src/import/ImportWavPack.cpp:241]: Ditto

**Review by crsib — CHANGES_REQUESTED** [review]
Decoding loop need some love :-)

And some minor are changes will be welcome!

File: src/export/ExportWavPack.cpp [hunk 17]
```diff
@@ -0,0 +1,502 @@
...
+            for (decltype(samplesThisRun) j = 0; j < samplesThisRun; j++) {
+               for (size_t i = 0; i < numChannels; i++) {
+                  wavpackBuffer[j*numChannels + i] = *mixed++;
+               }
+            }
+         } else {
+            const float *mixed = reinterpret_cast<const float*>(mixer->GetBuffer());
+            for (decltype(samplesThisRun) j = 0; j < samplesThisRun; j++) {
+               for (size_t i = 0; i < numChannels; i++) {
+                  int64_t intValue = (*mixed++) * (1u << 31);
```
Comment (Subhra264) [inline src/export/ExportWavPack.cpp:391]: In the `ExportWavPack.cpp` file, I think there were some issues with float 32 bit export. Earlier, I just passed 32 bit int values for float sample export. So, I think while importing it was getting imported as 32 bit int (before adding special cases for float sample import as you suggested). Though it was ok while importing in audacity ( wtih the previous code before applying your suggestions for import), I tried to play the 32 bit exported wavpack file using the VLC media player and it sounded like corrupted. Now, when I converted the `mixed` data into `float` and then into `int32_t` while exporting as 32 bit float, both VLC media and audacity (After applying your suggestions) opened and played the 32 bit float files correctly.

Comment (crsib) [reply src/export/ExportWavPack.cpp:391]: I think you should set `config->float_norm_exp: = 127;` to enable FP mode in Wavpack. In any case - you should consult `WavpackGetMode` to check, if you need to convert the values.

`1u << 31` - `std::numeric_limits<int32_t>::max()` please.

Comment (crsib) [reply src/export/ExportWavPack.cpp:391]: (Value `127` is an exponent  bias for IEEE754 32-bit float values. We have 8 bits for exponents, so the bias is `2^7 - 1`)

**Review by crsib — CHANGES_REQUESTED** [review]
Great work! Just a few more issue, but I think that they are easy to resolve.

File: src/export/ExportWavPack.cpp [hunk 18]
```diff
@@ -388,7 +389,7 @@ ProgressResult ExportWavPack::Export(AudacityProject *project,
             const float *mixed = reinterpret_cast<const float*>(mixer->GetBuffer());
             for (decltype(samplesThisRun) j = 0; j < samplesThisRun; j++) {
                for (size_t i = 0; i < numChannels; i++) {
-                  int64_t intValue = (*mixed++) * (1u << 31);
+                  int64_t intValue = (*mixed++) * (std::numeric_limits<int32_t>::max());
```
Comment (crsib) [inline src/export/ExportWavPack.cpp:392]: Minor, but I expect a compiler warning here. Let's add `static_cast<int64_t>` here. Sorry I've missed that before...

File: src/import/ImportWavPack.cpp [hunk 19]
```diff
@@ -0,0 +1,298 @@
...
+   {
+      const uint32_t bufferSize = mNumChannels * SAMPLES_TO_READ;
+      ArrayOf<int32_t> wavpackBuffer{ bufferSize };
+      uint32_t samplesRead = 0;
+
+      do {
+         samplesRead = WavpackUnpackSamples(mWavPackContext, wavpackBuffer.get(), SAMPLES_TO_READ);
+
+         if (mFormat == int16Sample) {
+            ArrayOf<int16_t> tempBuffer{ bufferSize };
```
Comment (crsib) [inline src/import/ImportWavPack.cpp:187]: Not that important here, but is is best to avoid memory allocations in such loops. Allocations can be quite costly.

**Review by crsib — CHANGES_REQUESTED** [review]
A one very minor step left!

**Review by crsib — APPROVED** [review]

File: src/export/ExportWavPack.cpp (hunk 14 above)
Comment (wthrajat) [reply src/export/ExportWavPack.cpp]: @crsib I am curious to know why :D
